File: packages/core/src/engine/verbs/unfold.ts
/*!
 * Copyright (c) Microsoft. All rights reserved.
 * Licensed under the MIT license. See LICENSE file in the project.
 */

import { op } from 'arquero'
import type ColumnTable from 'arquero/dist/types/table/column-table'
import type { TableStore, UnfoldArgs } from '../../index.js'
import type { Step } from '../../types.js'

/**
 * Executes an arquero fold operation. This creates two new columns:
 * one with the column name as key, the other with the row value.
 * @param step
 * @param store
 * @returns
 */
export async function unfold(
	step: Step,
	store: TableStore,
): Promise<ColumnTable> {
	const { input, args } = step
	const { key, value } = args as UnfoldArgs
	let inputTable = await store.table(input)

	inputTable = inputTable
		.groupby(key)
		.rollup({
			[value]: op.array_agg(value),
		})
		.pivot(key, value)

	return inputTable.unroll(inputTable.columnNames())
}

File: python/data-wrangling-components/data_wrangling_components/engine/verbs/unfold.py
#
# Copyright (c) Microsoft. All rights reserved.
# Licensed under the MIT license. See LICENSE file in the project.
#

from collections import defaultdict
from typing import Dict

import pandas as pd

from data_wrangling_components.table_store import TableContainer


def unfold(input: TableContainer, key: str, value: str):

    input_table = input.table

    output = input_table.copy()

    new_index = []
    id_key: Dict[str, int] = defaultdict(int)

    for temp_key in output[key]:
        new_index.append(id_key[temp_key])
        id_key[temp_key] += 1

    output.index = new_index

    output_temp = output.pivot(columns=key, values=value)
    other_columns = [column for column in output.columns if column not in [key, value]]
    output = pd.concat(
        [output[other_columns].groupby(level=0).agg("first"), output_temp], axis=1
    )

    return TableContainer(table=output)
